akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -63,69 +143,66 @@ private[akka] class AffinityPool(
   // indicates the current state of the pool
   @volatile final private var poolState: PoolState = Running
 
-  private final val workQueues = Array.fill(parallelism)(new BoundedTaskQueue(affinityGroupSize))
-  private final val workers = mutable.Set[ThreadPoolWorker]()
+  private final val workQueues = Array.fill(parallelism)(new BoundedAffinityTaskQueue(affinityGroupSize))
+  private final val workers = mutable.Set[AffinityPoolWorker]()
 
-  // a counter that gets incremented every time a task is queued
-  private val executionCounter: AtomicInteger = new AtomicInteger(0)
   // maps a runnable to an index of a worker queue
-  private val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
+  private[this] final val runnableToWorkerQueueIndex = new AtomicReference(ImmutableIntMap.empty)
 
-  private def locked[T](l: Lock)(body: ⇒ T) = {
-    l.lock()
+  @inline private def locked[T](body: ⇒ T): T = {
+    bookKeepingLock.lock()
     try {
       body
     } finally {
-      l.unlock()
+      bookKeepingLock.unlock()
     }
   }
 
-  private def getQueueForRunnable(command: Runnable) = {
-
+  private def getQueueForRunnable(command: Runnable): BoundedAffinityTaskQueue = {
     val runnableHash = command.hashCode()
 
-    def sbhash(i: Int) = reverseBytes(i * 0x9e3775cd) * 0x9e3775cd
-
-    def getNext = executionCounter.incrementAndGet() % parallelism
+    def indexFor(h: Int): Int =
+      Math.abs(reverseBytes(h * 0x9e3775cd) * 0x9e3775cd) % parallelism // In memory of Phil Bagwell
```

Comment:
In short: it's a hash improver such that consecutive hashCodes do not end up in consecutive slots (which is prevalent for Ints because the hashCode of 1 is 1, 2 is 2, etc.